### PR TITLE
SNLI task support 

### PR DESCRIPTION
**Pull request by vyskoto4**
Added SNLI task "task" and "preprocessing" and "loader" code. Simple README also included.

### CODE REVIEW

File: pysts/eval.py [hunk 1]
```diff
@@ -40,6 +40,42 @@ def binclass_accuracy(y, ypred):
 
     return (rawacc, y0acc, y1acc, balacc, f_score)
 
+def multiclass_accuracy(y, ypred):
+    """
+    Compute accuracy for multiclass classification tasks
+    Returns (rawacc, class_acc) where balacc is average of y0acc
+    and y1acc, regardless of their true balance in the dataset.
+    (The idea is that even if the unfortunate reality is that we have much
+    less y1 samples, their detection is equally important.)
```
Comment (pasky) [inline pysts/eval.py]: Rather describe class_acc than irrelevant balacc?

File: pysts/eval.py [hunk 2]
```diff
@@ -40,6 +40,42 @@ def binclass_accuracy(y, ypred):
...
+
+
+def eval_snli(ypred, y, name):
+    cls_names= ['contradiction', 'neutral', 'entailment']
+    rawacc,cls_acc = multiclass_accuracy(y, ypred)
+    print('%s Accuracy: %f \n%s accuracy %f\n%s accuracy %f\n%s accuracy %f' %(name, rawacc,
+                                                                                                             cls_names[0], cls_acc[0],
+                                                                                                             cls_names[1],cls_acc[1],
+                                                                                                             cls_names[2],cls_acc[2]))
+    return (rawacc)
```
Comment (pasky) [inline pysts/eval.py]: Have you considered printing everything on a single line, probably with just %.3f? Could that be more readable?

File: pysts/loader.py [hunk 3]
```diff
@@ -172,6 +173,35 @@ def load_sts(dsfile, skip_unlabeled=True):
...
+                continue
+            label = d['gold_label']
+            if label in lmappings:
+                s0 = word_tokenize(d['sentence1'])
+                s1 = word_tokenize(d['sentence2'])
+                s0i.append(s0)
+                s1i.append(s1)
+                labels.append(lmappings[label])
+            else:
+                neutral_skips+=1
```
Comment (pasky) [inline pysts/loader.py]: Does that happen? Might deserve a comment..

File: data/rte/snli/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+(on ``snli`` task).
+
+Model Comparison
+----------------
+
+For randomized models, 95% confidence intervals (t-distribution) will be reported.
+
+Note that NO DROPOUT is applied for any of the models.
+
+| Model                    | train    | test     | t. mean  | settings
```
Comment (pasky) [inline data/rte/snli/README.md]: Do you agree that we should also definitely use a validation set? If it's not explicitly published, what do the other papers do re validation/dev set?

File: tools/snli_preprocess.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    fraction=float(fraction)
+    print('Preparing train subset')
+    samplenum=np.ceil(fraction*TRAIN_LINES)
+    sample_inds=set(np.random.randint(TRAIN_LINES, size=samplenum))
+    build_subset(trainf, train_out, sample_inds)
+    # prepare test dataset
+    print('Preparing test subset')
+    samplenum=np.ceil(fraction*TEST_LINES)
+    sample_inds=set(np.random.randint(TEST_LINES, size=samplenum))
+    build_subset(testf, test_out, sample_inds)
```
Comment (pasky) [inline tools/snli_preprocess.py]: This isn't used anywhere, and it's a bit confusing...

**Comment by pasky** [comment]
Thanks, really awesome work!

I have added some comments, most of them are trivial nitpicking.  The important question we should address before merging is what to do re validation/test sets, probably by taking a look at what other papers do.

File: pysts/loader.py (hunk 3 above)
Comment (vyskoto4) [inline pysts/loader.py]: Yes it does happen. After the authors collected the dataset, they performed an additional round of data validation on a subsample. If 3 of 5 annotators agreed on one label, the sentence pair was assigned that label. If there was no such consensus, the pair was given a label ''-".

Will add that to comment in code.

Comment (pasky) [inline pysts/loader.py]: Thanks!

File: data/rte/snli/README.md (hunk 4 above)
Comment (vyskoto4) [inline data/rte/snli/README.md]: Dev set is now being used during training.

File: tools/train.py [hunk 6]
```diff
@@ -132,7 +132,8 @@ def train_and_eval(runid, module_prep_model, task, c, do_eval=True):
     print('Dataset')
     if 'vocabf' in conf:
         task.load_vocab(conf['vocabf'])
-    task.load_data(trainf, valf)
+    if 'testf' in conf:
+        task.load_data(trainf, valf,tesf)
```
Comment (pasky) [inline tools/train.py]: Hmm, a bit confused by this.  What's the purpose of this?  It even has a typo, so probably this change never actually comes into effect?

File: data/rte/snli/README.md (hunk 4 above)
Comment (pasky) [inline data/rte/snli/README.md]: Awesome! Waiting for results with train+dev for training and eval'd on test too, then I think we are clear to merge!

**Comment by pasky** [comment]
Tomas, I'll be away until Monday but it'd be really great if you'd have time to address the detail re the validation set until then so it's good to merge and we can include this in the paper.

**Comment by pasky** [comment]
Thanks!!

I have extended this a bit more.  Most importantly, I have changed the way accuracy is computed, to actually accuracy (I hope - comments welcome).  Please evaluate with the current master.

Great job otherwise. (There were some details that should deserve some attention, esp. very inconsistent whitespaces, but that's no big deal. :)
